crates/circuit/src/circuit_data.rs

```diff
@@ -2513,6 +2843,45 @@ impl CircuitData {
                                 // https://github.com/Qiskit/qiskit/issues/13504
                                 previous.py_op.take();
                             }
+                        } else if let OperationRef::ControlFlow(op) = previous_op.view() {
+                            let blocks = self.data[instruction].blocks_view();
+                            let block_to_edit = match op {
+                                ControlFlow::BreakLoop { .. } => Err(inconsistent()),
+                                ControlFlow::ContinueLoop { .. } => Err(inconsistent()),
+                                ControlFlow::ForLoop { .. } => {
+                                    match parameter {
+                                        2 => {
+                                            // In Python land, the loop body exists at parameter
+                                            // position 2.
+                                            Ok(blocks[0])
+                                        }
+                                        _ => Err(inconsistent()),
+                                    }
+                                }
+                                _ => {
+                                    // Most control flow instructions use the parameters for
+                                    // *just* their blocks.
+                                    Ok(blocks[parameter])
+                                }
+                            }?;
+                            if !seen_blocks.contains(&block_to_edit) {
+                                let mut new_block =
+                                    self.blocks[block_to_edit.index()].as_ref().clone();
```

Comment:
Note to future me: I think I probably want to use [Arc::make_mut](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.make_mut) here?